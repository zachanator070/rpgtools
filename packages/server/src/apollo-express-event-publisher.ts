import { EventPublisher } from "./types";
import { PubSub } from "apollo-server-express";
import { injectable } from "inversify";

@injectable()
export class ApolloExpressEventPublisher implements EventPublisher {
	pubsub = new PubSub();

	asyncIterator(events: string[]): AsyncIterator<any> {
		return this.pubsub.asyncIterator(events);
	}

	publish(event: string, payload: any): Promise<void> {
		return this.pubsub.publish(event, payload);
	}
}
