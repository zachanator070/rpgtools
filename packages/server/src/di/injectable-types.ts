import {DatabaseContext} from "../dal/database-context";

export const INJECTABLE_TYPES = {

	DomainEntity: Symbol.for("DomainEntity"),

	// domain entities
	Article: Symbol.for("Article"),
	Chunk: Symbol.for("Chunk"),
	File: Symbol.for("File"),
	Game: Symbol.for("Game"),
	Image: Symbol.for("Image"),
	Item: Symbol.for("Item"),
	Model: Symbol.for("Model"),
	Monster: Symbol.for("Monster"),
	PermissionAssignment: Symbol.for("PermissionAssignment"),
	Person: Symbol.for("Person"),
	Pin: Symbol.for("Pin"),
	Place: Symbol.for("Place"),
	Role: Symbol.for("Role"),
	ServerConfig: Symbol.for("ServerConfig"),
	User: Symbol.for("User"),
	WikiFolder: Symbol.for("WikiFolder"),
	World: Symbol.for("World"),
	Event: Symbol.for("Event"),
	Calendar: Symbol.for("Calendar"),

	// entity factories
	ArticleFactory: Symbol.for("ArticleFactory"),
	AclFactory: Symbol.for('AclFactory'),
	ChunkFactory: Symbol.for("ChunkFactory"),
	FileFactory: Symbol.for("FileFactory"),
	GameFactory: Symbol.for("GameFactory"),
	ImageFactory: Symbol.for("ImageFactory"),
	ItemFactory: Symbol.for("ItemFactory"),
	ModelFactory: Symbol.for("ModelFactory"),
	MonsterFactory: Symbol.for("MonsterFactory"),
	PersonFactory: Symbol.for("PersonFactory"),
	PinFactory: Symbol.for("PinFactory"),
	PlaceFactory: Symbol.for("PlaceFactory"),
	RoleFactory: Symbol.for("RoleFactory"),
	ServerConfigFactory: Symbol.for("ServerConfigFactory"),
	UserFactory: Symbol.for("UserFactory"),
	WikiFolderFactory: Symbol.for("WikiFolderFactory"),
	WikiPageFactory: Symbol.for("WikiPageFactory"),
	WorldFactory: Symbol.for("WorldFactory"),
	EventWikiFactory: Symbol.for("EventFactory"),
	CalendarFactory: Symbol.for("CalendarFactory"),

	CharacterFactory: Symbol.for('CharacterFactory'),
	CharacterAttributeFactory: Symbol.for('CharacterAttributeFactory'),
	PathNodeFactory: Symbol.for('NodePathFactory'),
	StrokeFactory: Symbol.for('StrokeFactory'),
	FogStrokeFactory: Symbol.for('FogStrokeFactory'),
	MessageFactory: Symbol.for('MessageFactory'),
	InGameModelFactory: Symbol.for('InGameModelFactory'),

	AgeFactory: Symbol.for("AgeFactory"),
	MonthFactory: Symbol.for("MonthFactory"),
	DayOfTheWeekFactory: Symbol.for("DayOfTheWeekFactory"),

	// repositories
	ArticleRepository: Symbol.for("ArticleRepository"),
	ChunkRepository: Symbol.for("ChunkRepository"),
	FileRepository: Symbol.for("FileRepository"),
	GameRepository: Symbol.for("GameRepository"),
	ImageRepository: Symbol.for("ImageRepository"),
	ItemRepository: Symbol.for("ItemRepository"),
	ModelRepository: Symbol.for("ModelRepository"),
	MonsterRepository: Symbol.for("MonsterRepository"),
	PersonRepository: Symbol.for("PersonRepository"),
	PinRepository: Symbol.for("PinRepository"),
	PlaceRepository: Symbol.for("PlaceRepository"),
	RoleRepository: Symbol.for("RoleRepository"),
	ServerConfigRepository: Symbol.for("ServerConfigRepository"),
	UserRepository: Symbol.for("UserRepository"),
	WikiFolderRepository: Symbol.for("WikiFolderRepository"),
	WikiPageRepository: Symbol.for("WikiPageRepository"),
	WorldRepository: Symbol.for("WorldRepository"),
	EventWikiRepository: Symbol.for("EventWikiRepository"),
	CalendarRepository: Symbol.for("CalendarRepository"),
	StrokeRepository: Symbol.for('StrokeRepository'),
	FogStrokeRepository: Symbol.for('FogStrokeRepository'),

	SqlPermissionControlledRepository: Symbol.for('SqlPermissionControlledRepository'),

	// authorization policies
	ArticleAuthorizationPolicy: Symbol.for("ArticleAuthorizationPolicy"),
	ChunkAuthorizationPolicy: Symbol.for("ChunkAuthorizationPolicy"),
	FileAuthorizationPolicy: Symbol.for("FileAuthorizationPolicy"),
	GameAuthorizationPolicy: Symbol.for("GameAuthorizationPolicy"),
	ImageAuthorizationPolicy: Symbol.for("ImageAuthorizationPolicy"),
	ItemAuthorizationPolicy: Symbol.for("ItemAuthorizationPolicy"),
	ModelAuthorizationPolicy: Symbol.for("ModelAuthorizationPolicy"),
	MonsterAuthorizationPolicy: Symbol.for("MonsterAuthorizationPolicy"),
	PersonAuthorizationPolicy: Symbol.for("PersonAuthorizationPolicy"),
	PinAuthorizationPolicy: Symbol.for("PinAuthorizationPolicy"),
	PlaceAuthorizationPolicy: Symbol.for("PlaceAuthorizationPolicy"),
	RoleAuthorizationPolicy: Symbol.for("RoleAuthorizationPolicy"),
	ServerConfigAuthorizationPolicy: Symbol.for("ServerConfigAuthorizationPolicy"),
	UserAuthorizationPolicy: Symbol.for("UserAuthorizationPolicy"),
	WikiFolderAuthorizationPolicy: Symbol.for("WikiFolderAuthorizationPolicy"),
	WikiPageAuthorizationPolicy: Symbol.for("WikiPageAuthorizationPolicy"),
	WorldAuthorizationPolicy: Symbol.for("WorldAuthorizationPolicy"),
	CalendarAuthorizationPolicy: Symbol.for("CalendarAuthorizationPolicy"),
	FogStrokeAuthorizationPolicy: Symbol.for('FogStrokeAuthorizationPolicy'),
	StrokeAuthorizationPolicy: Symbol.for('StrokeAuthorizationPolicy'),

	// archive repositories
	ArchiveArticleRepository: Symbol.for("ArchiveArticleRepository"),
	ArchiveChunkRepository: Symbol.for("ArchiveChunkRepository"),
	ArchiveGameRepository: Symbol.for("ArchiveGameRepository"),
	ArchiveImageRepository: Symbol.for("ArchiveImageRepository"),
	ArchiveItemRepository: Symbol.for("ArchiveItemRepository"),
	ArchiveModelRepository: Symbol.for("ArchiveModelRepository"),
	ArchiveMonsterRepository: Symbol.for("ArchiveMonsterRepository"),
	ArchivePermissionAssignmentRepository: Symbol.for("ArchivePermissionAssignmentRepository"),
	ArchivePersonRepository: Symbol.for("ArchivePersonRepository"),
	ArchivePinRepository: Symbol.for("ArchivePinRepository"),
	ArchivePlaceRepository: Symbol.for("ArchivePlaceRepository"),
	ArchiveRoleRepository: Symbol.for("ArchiveRoleRepository"),
	ArchiveServerConfigRepository: Symbol.for("ArchiveServerConfigRepository"),
	ArchiveUserRepository: Symbol.for("ArchiveUserRepository"),
	ArchiveWikiFolderRepository: Symbol.for("ArchiveWikiFolderRepository"),
	ArchiveWikiPageRepository: Symbol.for("ArchiveWikiPageRepository"),
	ArchiveWorldRepository: Symbol.for("ArchiveWorldRepository"),
	ArchiveFileRepository: Symbol.for("ArchiveFileRepository"),

	// services
	AuthenticationService: Symbol.for("AuthenticationService"),
	AuthorizationService: Symbol.for("AuthorizationService"),
	ContentExportService: Symbol.for("ContentExportService"),
	ContentImportService: Symbol.for("ContentImportService"),
	GameService: Symbol.for("GameService"),
	ImageService: Symbol.for("ImageService"),
	ModelService: Symbol.for("ModelService"),
	RoleService: Symbol.for("RoleService"),
	ServerConfigService: Symbol.for("ServerConfigService"),
	SrdImportService: Symbol.for("SrdImportService"),
	UserService: Symbol.for("UserService"),
	WikiFolderService: Symbol.for("WikiFolderService"),
	WikiPageService: Symbol.for("WikiPageService"),
	WorldService: Symbol.for("WorldService"),

	// session contexts
	SessionContextFactory: Symbol.for("SessionContextFactory"),
	SecurityContextFactory: Symbol.for("SecurityContextFactory"),

	// 5e
	Open5eApiClient: Symbol.for("Open5eApiClient"),
	Dnd5eApiClient: Symbol.for("Dnd5eApiClient"),

	// delta
	DeltaFactory: Symbol.for("DeltaFactory"),

	// cache
	Cache: Symbol.for("Cache"),

	// archive
	ArchiveFactory: Symbol.for("ArchiveFactory"),

	ZipArchiveFactory: Symbol.for("ZipArchiveFactory"),
	ZipArchive: Symbol.for('ZipArchive'),

	// event publisher
	EventPublisher: Symbol.for("EventPublisher"),

	// server
	ApiServer: Symbol.for("ApiServer"),
	ServerProperties: Symbol.for("ServerProperties"),
	RpgToolsServer: Symbol.for('RpgToolsServer'),
	DbEngine: Symbol.for("DbEngine"),
	SqlDbEngine: Symbol.for('SqlDbEngine'),

	// data loaders
	ArticleDataLoader: Symbol.for("ArticleDataLoader"),
	ChunkDataLoader: Symbol.for("ChunkDataLoader"),
	FileDataLoader: Symbol.for("FileDataLoader"),
	GameDataLoader: Symbol.for("GameDataLoader"),
	ImageDataLoader: Symbol.for("ImageDataLoader"),
	ItemDataLoader: Symbol.for("ItemDataLoader"),
	ModelDataLoader: Symbol.for("ModelDataLoader"),
	MonsterDataLoader: Symbol.for("MonsterDataLoader"),
	PermissionAssignmentDataLoader: Symbol.for("PermissionAssignmentDataLoader"),
	PersonDataLoader: Symbol.for("PersonDataLoader"),
	PinDataLoader: Symbol.for("PinDataLoader"),
	PlaceDataLoader: Symbol.for("PlaceDataLoader"),
	RoleDataLoader: Symbol.for("RoleDataLoader"),
	ServerConfigDataLoader: Symbol.for("ServerConfigDataLoader"),
	UserDataLoader: Symbol.for("UserDataLoader"),
	WikiFolderDataLoader: Symbol.for("WikiFolderDataLoader"),
	WikiPageDataLoader: Symbol.for("WikiPageDataLoader"),
	WorldDataLoader: Symbol.for("WorldDataLoader"),
	CalendarDataLoader: Symbol.for('CalendarDataLoader'),
	EventDataLoader: Symbol.for('EventDataLoader'),

	// seeders
	RoleSeeder: Symbol.for("RoleSeeder"),
	ServerConfigSeeder: Symbol.for("ServerConfigSeeder"),

	// database context
	DatabaseContext: Symbol.for("DatabaseContext"),
	DatabaseContextFactory: Symbol.for("DatabaseContextFactory"),

	// repo mapper
	EntityMapper: Symbol.for("EntityMapper"),

	// mongodb
	FilterFactory: Symbol.for("FilterFactory"),
	MongoDbMigrationV40: Symbol.for("MongoDbMigrationV40"),


};
