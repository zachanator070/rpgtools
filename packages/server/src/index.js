import express from "express";
import bodyParser from "body-parser";
import gqlServer from "./gql-server";
import cookieParser from "cookie-parser";
import morgan from "morgan";
import mongoose from "mongoose";
import { defaultSeeders } from "./default-seeders";
import path from "path";
import { ImageRouter } from "./routers/image-router";
import { checkConfig, serverNeedsSetup } from "./server-needs-setup";
import crypto from "crypto";
import * as http from "http";
import { ServerConfig } from "./models/server-config";
import { ModelRouter } from "./routers/model-router";
import ExportRouter from "./routers/export-router";
import { graphqlUploadExpress } from "graphql-upload";

const mongodb_host = process.env.MONGODB_HOST || "mongodb";
const mongodb_db_name = process.env.MONGODB_DB_NAME || "rpgtools";

mongoose
  .connect(`mongodb://${mongodb_host}/${mongodb_db_name}`, {
    useNewUrlParser: true,
  })
  .then(async () => {
    console.log(
      `Connected to mongodb at mongodb://${mongodb_host}/${mongodb_db_name}`
    );
    await defaultSeeders();
    await createServer();
  });

mongoose.set("useNewUrlParser", true);
mongoose.set("useFindAndModify", false);
mongoose.set("useCreateIndex", true);
mongoose.set("useUnifiedTopology", true);

const createServer = async () => {
  const server = express();

  server.use(bodyParser.json());
  server.use(morgan("tiny"));

  server.use(cookieParser());

  await checkConfig();
  const serverConfig = await ServerConfig.findOne();
  if (serverNeedsSetup()) {
    console.warn(
      `Server needs configuration! Use unlock code ${serverConfig.unlockCode} to unlock`
    );
  }

  server.get("*.js", function (req, res, next) {
    req.url = req.url + ".gz";
    res.set("Content-Encoding", "gzip");
    res.set("Content-Type", "text/javascript");
    next();
  });
  server.get("*.css", function (req, res, next) {
    req.url = req.url + ".gz";
    res.set("Content-Encoding", "gzip");
    res.set("Content-Type", "text/css");
    next();
  });
  server.get("*favicon.ico", function (req, res, next) {
    req.url = "/favicon.ico";
    next();
  });

  server.use("/images", ImageRouter);
  server.use("/models", ModelRouter);
  server.use("/export", ExportRouter);

  server.get("*", (req, res, next) => {
    const needsSetup = serverNeedsSetup();
    if (needsSetup) {
      if (
        [
          "/api",
          "/ui/setup",
          "/app.bundle.js",
          "/app.css",
          "/app.bundle.js.gz",
          "/app.css.gz",
          "favicon.ico",
        ].includes(req.url)
      ) {
        return next();
      }
      return res.redirect(302, "/ui/setup");
    } else {
      return next();
    }
  });

  server.get("/ui*", (req, res) => {
    return res.sendFile(path.resolve(__dirname, "../../dist", "index.html"));
  });

  server.use(express.static("dist"));
  server.use(express.static("src/static-assets"));
  server.use(graphqlUploadExpress());

  gqlServer.applyMiddleware({ app: server, path: "/api" });

  if (!process.env.ACCESS_TOKEN_SECRET) {
    console.warn(
      "environment variable ACCESS_TOKEN_SECRET is not set, restarting server will log out all users"
    );
    process.env.ACCESS_TOKEN_SECRET = crypto.randomBytes(2048);
  }

  if (!process.env.REFRESH_TOKEN_SECRET) {
    console.log(
      "environment variable REFRESH_TOKEN_SECRET is not set, restarting server will log out all users"
    );
    process.env.REFRESH_TOKEN_SECRET = crypto.randomBytes(2048);
  }

  const port = process.env.SERVER_PORT || 3000;

  const httpServer = http.createServer(server);
  gqlServer.installSubscriptionHandlers(httpServer);

  httpServer.listen(port, () => {
    console.log(
      `The server is running and listening at http://localhost:${port}`
    );
  });
};
