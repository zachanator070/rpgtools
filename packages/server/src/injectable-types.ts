export const INJECTABLE_TYPES = {
	ArticleRepository: Symbol.for("ArticleRepository"),
	ChunkRepository: Symbol.for("ChunkRepository"),
	GameRepository: Symbol.for("GameRepository"),
	ImageRepository: Symbol.for("ImageRepository"),
	ItemRepository: Symbol.for("ItemRepository"),
	ModelRepository: Symbol.for("ModelRepository"),
	MonsterRepository: Symbol.for("MonsterRepository"),
	PermissionAssignmentRepository: Symbol.for("PermissionAssignmentRepository"),
	PersonRepository: Symbol.for("PersonRepository"),
	PinRepository: Symbol.for("PinRepository"),
	PlaceRepository: Symbol.for("PlaceRepository"),
	RoleRepository: Symbol.for("RoleRepository"),
	ServerConfigRepository: Symbol.for("ServerConfigRepository"),
	UserRepository: Symbol.for("UserRepository"),
	WikiFolderRepository: Symbol.for("WikiFolderRepository"),
	WikiPageRepository: Symbol.for("WikiPageRepository"),
	WorldRepository: Symbol.for("WorldRepository"),
	FileRepository: Symbol.for("FileRepository"),

	AuthenticationService: Symbol.for("AuthenticationService"),
	AuthorizationService: Symbol.for("AuthorizationService"),
	ContentExportService: Symbol.for("ContentExportService"),
	ContentImportService: Symbol.for("ContentImportService"),
	GameService: Symbol.for("GameService"),
	ImageService: Symbol.for("ImageService"),
	ModelService: Symbol.for("ModelService"),
	ServerConfigService: Symbol.for("ServerConfigService"),
	SrdImportService: Symbol.for("SrdImportService"),
	UserService: Symbol.for("UserService"),
	WikiFolderService: Symbol.for("WikiFolderService"),
	WikiPageService: Symbol.for("WikiPageService"),
	WorldService: Symbol.for("WorldService"),

	SessionContextFactory: Symbol.for("SessionContextFactory"),
	Open5eApiClient: Symbol.for("Open5eApiClient"),

	Cache: Symbol.for("Cache"),
};
