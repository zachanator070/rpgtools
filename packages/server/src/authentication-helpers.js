import jwt from "jsonwebtoken";
import { v4 as uuidv4 } from "uuid";
import { User } from "./models/user";
import { ANON_USERNAME } from "@rpgtools/common/src/permission-constants";

export const authenticated = (next) => (root, args, context, info) => {
  if (!context.currentUser || context.currentUser.username === ANON_USERNAME) {
    throw new Error(`Authentication is required for this action`);
  }

  return next(root, args, context, info);
};

export const createTokens = async (user) => {
  const accessToken = jwt.sign(
    { userId: user._id },
    process.env["ACCESS_TOKEN_SECRET"],
    { expiresIn: ACCESS_TOKEN_MAX_AGE.string }
  );
  let version = uuidv4();
  const refreshToken = jwt.sign(
    { version: version, userId: user._id },
    process.env["REFRESH_TOKEN_SECRET"],
    { expiresIn: REFRESH_TOKEN_MAX_AGE.string }
  );
  user.tokenVersion = version;
  await user.save();
  return { accessToken, refreshToken };
};

export const getCurrentUser = async (userId) => {
  return User.findById(userId)
    .populate({ path: "currentWorld", populate: { path: "wikiPage" } })
    .populate({ path: "permissions", populate: "subject" })
    .populate({ path: "roles", populate: "permissions" });
};

export const createSessionContext = async ({ req, res, connection }) => {
  let currentUser = null;
  if (connection) {
    return connection.context;
  }

  const refreshToken = req.cookies["refreshToken"];
  const accessToken = req.cookies["accessToken"];

  // if access token wasn't given but refresh token was
  if (accessToken || refreshToken) {
    try {
      let data = jwt.verify(accessToken, process.env["ACCESS_TOKEN_SECRET"], {
        maxAge: ACCESS_TOKEN_MAX_AGE.string,
      });
      currentUser = await getCurrentUser(data.userId);
    } catch (e) {
      // accessToken is expired
      console.log(`Access token expired because ${e.message}`);
      res.clearCookie("accessToken");
      try {
        let data = jwt.verify(
          refreshToken,
          process.env["REFRESH_TOKEN_SECRET"],
          { maxAge: REFRESH_TOKEN_MAX_AGE.string }
        );
        currentUser = await getCurrentUser(data.userId);
        // if refreshToken is still valid issue new access token and refresh token
        if (currentUser && currentUser.tokenVersion === data.version) {
          console.log(
            "Access token and refresh token renewed b/c refresh token still valid"
          );
          let tokens = await createTokens(currentUser);
          res.cookie("accessToken", tokens.accessToken, {
            maxAge: ACCESS_TOKEN_MAX_AGE.ms,
            secure: false,
          });
          res.cookie("refreshToken", tokens.refreshToken, {
            maxAge: REFRESH_TOKEN_MAX_AGE.ms,
            secure: false,
          });
        } else {
          // refreshToken was invalidated
          console.log(
            `Refresh token expired because version did not match ${currentUser.tokenVersion} !== ${data.version}`
          );
          currentUser = null;
          res.clearCookie("refreshToken");
        }
      } catch (e) {
        // refreshToken is expired
        console.log(`Refresh token expired because ${e.message}`);
        throw new Error("Refresh token expired");
      }
    }
  }

  if (!currentUser) {
    currentUser = new User({ username: ANON_USERNAME });
  }

  await currentUser.recalculateAllPermissions();

  return {
    res,
    currentUser,
  };
};

// export const ACCESS_TOKEN_MAX_AGE = {string: '15min', ms: 1000 * 60 * 15};
export const ACCESS_TOKEN_MAX_AGE = { string: "1m", ms: 1000 * 60 };
export const REFRESH_TOKEN_MAX_AGE = { string: "1d", ms: 1000 * 60 * 60 * 24 };
