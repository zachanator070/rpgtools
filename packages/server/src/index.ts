import {container} from "./di/inversify";
import {INJECTABLE_TYPES} from "./di/injectable-types";
import {ApiServer} from "./types";

const server = container.get<ApiServer>(INJECTABLE_TYPES.ApiServer);

(async () => {
	await server.start();
})();
