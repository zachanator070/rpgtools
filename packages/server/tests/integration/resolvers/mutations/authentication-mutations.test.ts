import { container } from "../../../../src/di/inversify";
import { ServerConfigRepository } from "../../../../src/types";
import { INJECTABLE_TYPES } from "../../../../src/di/injectable-types";
import { defaultTestingContextFactory } from "../../DefaultTestingContextFactory";
import {gql} from "graphql-tag";

process.env.TEST_SUITE = "authentication-mutations-test";

export const REGISTER_MUTATION = gql`
	mutation register(
		$registerCode: String!
		$email: String!
		$username: String!
		$password: String!
	) {
		register(
			registerCode: $registerCode
			email: $email
			username: $username
			password: $password
		) {
			_id
		}
	}
`;

export const LOGIN_QUERY = gql`
	mutation login($username: String!, $password: String!) {
		login(username: $username, password: $password) {
			_id
		}
	}
`;

describe("authentication-mutations", () => {
	let {
		server,
		mockSessionContextFactory,
		otherUser,
		otherUserSecurityContext,
		world,
		testRole,
		currentUser,
		testerSecurityContext,
		newFolder,
		...testingContext
	} = defaultTestingContextFactory();

	test("login", async () => {
		const result = await server.executeGraphQLQuery({
			query: LOGIN_QUERY,
			variables: { username: "tester", password: "tester" },
		});
		expect(result).toMatchSnapshot({
			data: {
				login: {
					_id: expect.any(String),
				},
			},
			errors: undefined,
		});
	});

	test("login bad password", async () => {
		const result = await server.executeGraphQLQuery({
			query: LOGIN_QUERY,
			variables: { username: "tester", password: "asdf" },
		});
		expect(result).toMatchSnapshot();
	});

	test("login bad username", async () => {
		const result = await server.executeGraphQLQuery({
			query: LOGIN_QUERY,
			variables: { username: "tester", password: "asdf" },
		});
		expect(result).toMatchSnapshot();
	});

	describe("with good register code available", () => {
		beforeEach(async () => {
			const repo = container.get<ServerConfigRepository>(INJECTABLE_TYPES.ServerConfigRepository);
			const serverConfig = await repo.findOne();
			serverConfig.registerCodes.push("asdf");
			await repo.update(serverConfig);
		});

		test("register good", async () => {
			const result = await server.executeGraphQLQuery({
				query: REGISTER_MUTATION,
				variables: {
					registerCode: "asdf",
					email: "asdf@gmail.com",
					username: "tester2",
					password: "tester",
				},
			});
			expect(result).toMatchSnapshot({
				data: {
					register: {
						_id: expect.any(String),
					},
				},
				errors: undefined,
			});
		});

		test("register use code twice", async () => {
			await server.executeGraphQLQuery({
				query: REGISTER_MUTATION,
				variables: {
					registerCode: "asdf",
					email: "asdf@gmail.com",
					username: "tester2",
					password: "tester",
				},
			});
			const result = await server.executeGraphQLQuery({
				query: REGISTER_MUTATION,
				variables: {
					registerCode: "asdf",
					email: "asdf@gmail.com",
					username: "tester2",
					password: "tester",
				},
			});
			expect(result).toMatchSnapshot();
		});

		test("register use email twice", async () => {
			await server.executeGraphQLQuery({
				query: REGISTER_MUTATION,
				variables: {
					registerCode: "asdf",
					email: "asdf@gmail.com",
					username: "tester2",
					password: "tester",
				},
			});
			const result = await server.executeGraphQLQuery({
				query: REGISTER_MUTATION,
				variables: {
					registerCode: "qwerty",
					email: "asdf@gmail.com",
					username: "tester3",
					password: "tester",
				},
			});
			expect(result).toMatchSnapshot();
		});

		test("register use username twice", async () => {
			await server.executeGraphQLQuery({
				query: REGISTER_MUTATION,
				variables: {
					registerCode: "asdf",
					email: "tester2@gmail.com",
					username: "tester2",
					password: "tester",
				},
			});
			const result = await server.executeGraphQLQuery({
				query: REGISTER_MUTATION,
				variables: {
					registerCode: "qwerty",
					email: "tester3@gmail.com",
					username: "tester2",
					password: "tester",
				},
			});
			expect(result).toMatchSnapshot();
		});
	});

	test("register bad code", async () => {
		const result = await server.executeGraphQLQuery({
			query: REGISTER_MUTATION,
			variables: {
				registerCode: "1234",
				email: "asdf@gmail.com",
				username: "tester2",
				password: "tester",
			},
		});
		expect(result).toMatchSnapshot();
	});
});
