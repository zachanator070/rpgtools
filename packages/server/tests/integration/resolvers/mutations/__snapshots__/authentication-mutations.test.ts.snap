// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authentication-mutations login 1`] = `
Object {
  "data": Object {
    "login": Object {
      "_id": Any<String>,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authentication-mutations login bad password 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Login failure: username or password are incorrect],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authentication-mutations login bad username 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Login failure: username or password are incorrect],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authentication-mutations register bad code 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Register code not valid],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authentication-mutations with good register code available register good 1`] = `
Object {
  "data": Object {
    "register": Object {
      "_id": Any<String>,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authentication-mutations with good register code available register use code twice 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Register code not valid],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authentication-mutations with good register code available register use email twice 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Register code not valid],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authentication-mutations with good register code available register use username twice 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Register code not valid],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
